lib/internal/fs/promises.js

```diff
@@ -87,70 +97,91 @@ class FileHandle extends JSTransferable {
   }
 
   appendFile(data, options) {
-    return writeFile(this, data, options);
+    return fsCall(writeFile, this, data, options);
   }
 
   chmod(mode) {
-    return fchmod(this, mode);
+    return fsCall(fchmod, this, mode);
   }
 
   chown(uid, gid) {
-    return fchown(this, uid, gid);
+    return fsCall(fchown, this, uid, gid);
   }
 
   datasync() {
-    return fdatasync(this);
+    return fsCall(fdatasync, this);
   }
 
   sync() {
-    return fsync(this);
+    return fsCall(fsync, this);
   }
 
   read(buffer, offset, length, position) {
-    return read(this, buffer, offset, length, position);
+    return fsCall(read, this, buffer, offset, length, position);
   }
 
   readv(buffers, position) {
-    return readv(this, buffers, position);
+    return fsCall(readv, this, buffers, position);
   }
 
   readFile(options) {
-    return readFile(this, options);
+    return fsCall(readFile, this, options);
   }
 
   stat(options) {
-    return fstat(this, options);
+    return fsCall(fstat, this, options);
   }
 
   truncate(len = 0) {
-    return ftruncate(this, len);
+    return fsCall(ftruncate, this, len);
   }
 
   utimes(atime, mtime) {
-    return futimes(this, atime, mtime);
+    return fsCall(futimes, this, atime, mtime);
   }
 
   write(buffer, offset, length, position) {
-    return write(this, buffer, offset, length, position);
+    return fsCall(write, this, buffer, offset, length, position);
   }
 
   writev(buffers, position) {
-    return writev(this, buffers, position);
+    return fsCall(writev, this, buffers, position);
   }
 
   writeFile(data, options) {
-    return writeFile(this, data, options);
+    return fsCall(writeFile, this, data, options);
   }
 
   close = () => {
-    this[kFd] = -1;
-    return this[kHandle].close();
+    if (this[kClosePromise]) {
+      return this[kClosePromise];
+    }
+
+    this[kRefs]--;
+    if (this[kRefs] === 0) {
+      this[kFd] = -1;
+      this[kClosePromise] = this[kHandle].close();
+    } else {
+      this[kClosePromise] = new Promise((resolve, reject) => {
+        this[kCloseResolve] = resolve;
+        this[kCloseReject] = reject;
+      });
+    }
```

Comment:
```suggestion
    this[kRefs]--;
    if (this[kRefs] === 0) {
      this[kFd] = -1;
      this[kClosePromise] = this[kHandle].close().finally(() => {
        this[kClosePromise] = undefined;
      });
    } else {
      this[kClosePromise] = new Promise((resolve, reject) => {
        this[kCloseResolve] = resolve;
        this[kCloseReject] = reject;
      }).finally(() => {
        this[kClosePromise] = undefined;
        this[kCloseReject] = undefined;
        this[kCloseResolve] = undefined;
      });
    }
```

Release those references to avoid possible memory leaks...

Reply:
I don't see how that could leak? Also I think it's good to keep them in case of multiple calls to close().

Reply:
They leak because the references are kept around longer than they need to after the `close()` Promise resolves. Yes, once the `FileHandle` is gc'd they'll be released but it's possible that could take a while longer. Since we no longer need them, we should clean them up as soon as possible.

Multiple calls to `close()` should be fine. If the `FileHandle` is already closed (`fd === -1`)  simply `return Promise.resolve()` immediately at the start of `close()`.

Reply:
> Multiple calls to close() should be fine

I was more thinking along the lines if close errors and you call it again it will throw the same error again.

Reply:
> They leak because the references are kept around longer than they need to after the close() 

Not sure I follow. They will be kept around until GC anyway. Unless you mean that the FileHandle is in a different GC generation?

Reply:
Yes, that's it exactly.

Reply:
Cool, and what about:

> I was more thinking along the lines if close errors and you call it again it will throw the same error again.

Reply:
I doubt that's necessary. It's rare to get an error on close and if you do, crashing is likely the best choice... even if you choose to ignore it, you're likely going to want to be able to try again or risk leaking the underlying native `fd`. After the `FileHandle` is closed, making it a simple `return Promise.resolve()` non-op makes the most sense.